org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/codeActions/QuickFixCodeActionService.java

```diff
@@ -44,26 +46,28 @@ public class QuickFixCodeActionService implements ICodeActionService2 {
 
 	@Override
 	public List<Either<Command, CodeAction>> getCodeActions(Options options) {
-		List<Either<Command, CodeAction>> result = new ArrayList<>();
-
 		boolean handleQuickfixes = options.getCodeActionParams().getContext().getOnly() == null
 				|| options.getCodeActionParams().getContext().getOnly().isEmpty()
 				|| options.getCodeActionParams().getContext().getOnly().contains(CodeActionKind.QuickFix);
 
-		if (handleQuickfixes) {
-			for (Diagnostic error : options.getCodeActionParams().getContext().getDiagnostics()) {
-				List<DiagnosticResolution> resolutions = quickfixes.getResolutions(options, error);
-				for (DiagnosticResolution resolution : resolutions) {
-					result.add(Either.forRight(createFix(resolution, error)));
-				}
+		if (!handleQuickfixes) {
+			return Collections.emptyList();
+		}
+
+		List<Either<Command, CodeAction>> result = new ArrayList<>();
+		for (Diagnostic diagnostic : options.getCodeActionParams().getContext().getDiagnostics()) {
+			List<DiagnosticResolution> resolutions = quickfixes.getResolutions(options, diagnostic).stream()
+					.sorted(Comparator.comparing(DiagnosticResolution::getLabel)).collect(Collectors.toList());
```

Comment:
Not sure this comparator will work. I had a quick look at the uses of `getLabel` and I see some places null checking it. The datastructure `DiagnosticResolution` does not make sure that the label is not null and as far as I can see, `Comparator.comparing` can't handle keyExtractors that can produce nulls